week4/trie/jasonkuhrt/Main.hs

```diff
@@ -0,0 +1,234 @@
+
+     a and
+    be bean
+
+As a visualized trie:
+
+      a       b
+     /         \
+    n           e
+                 \
+                  a
+                  |
+                  n
+
+So our solution is to model each Node with a flag indicating whether it
+represents a valid word or not at that point in the path. For example, the
+above visualization in pseudo code:
+
+    Node False
+      a : Node True
+        n : Empty
+      b : Node False
+        e : Node True
+          a : Node False
+            n : Empty
+
+Finally each letter (a Node) may branch out to zero, one, or many other
+letters (nodes) which themselves finish or continue toward another word.
+For example:
+
+        b
+       /
+      e
+     /|\
+    t e  a
+      |  |
+      r  n
+
+Our solution is to use a Map data type which permits us to model arbitrary
+number of children per node.
+-}
+data Trie a =
+  Empty |
+  Node Bool (Map a (Trie a))
+  deriving (Show)
+
+sample :: Trie Char
+sample =
+  Node False (Map.fromList [
+    ('a', Node True (Map.fromList [
+      ('n', Empty)
+    ])),
+    ('b', Node False (Map.fromList [
+      ('e', Node True (Map.fromList [
+        ('a', Node False (Map.fromList [
+          ('n', Empty)
+        ]))
+      ]))
+    ]))
+  ])
+
+
+
+-- | Interactive search on the command line.
+main :: IO ()
+main = do
+  putStr "==> Indexing dictionary... "
+  trie <- pure . wordsToTrie . lines =<< readFile "/usr/share/dict/words"
+  putStrLn "Done!"
+  putStrLn "==> Now search for words by prefix."
+  prompt trie
+  where
+  prompt trie = do
+    putStr "> "
+    putStrLn . unwords . flip search trie =<< getLine
+    putStrLn ""
+    prompt trie
+
+
+
+{- Index a dictionary of words.
+
+For example:
+
+  > wordsToTrie ["bet", "be", "bean","beam"]
+
+  Node False (fromList [
+    ('b', Node False (fromList [
+      ('e', Node True (fromList [
+        ('t', Empty),
+        ('a', Node False (fromList [
+          ('m', Empty),
+          ('n',Empty)
+        ]))
+      ]))
+    ]))
+  ])
+-}
+wordsToTrie :: [String] -> Trie Char
+wordsToTrie = foldl indexWord (Node False Map.empty)
+
+indexWord :: Trie Char -> String -> Trie Char
+-- Finished indexing word whose entire path already existed
+indexWord (Node _ branches)  ""         = Node True branches
+-- Extend a trie path
+indexWord Empty     string              = stringToTrie True string
+indexWord (Node isWord maap) (char:cs)
+  -- Word path already exists
+  | Map.member char maap =
+    Node isWord (Map.update (Just . flip indexWord cs) char maap)
+  -- Word path is novel
+  | otherwise =
+    Node isWord (Map.insert char (stringToTrie False cs) maap)
+
+stringToTrie :: Bool -> String -> Trie Char
+stringToTrie _      ""     = Empty
+stringToTrie isWord (c:cs) = Node isWord (Map.singleton c (stringToTrie False cs))
+
+
+
+-- | Find words in given trie that have given prefix.
+search :: String -> Trie Char -> [String]
+search string = trieToWords . findTrieWithPrefix string
```

Comment:
Instead of this, it might be easier to lookup the *branch* of the trie containing words with a prefix, use `trieToWords` on it and then add the prefix to all words.